rlclientlib/extensions/onnx/src/tensor_notation.cc

```diff
@@ -0,0 +1,312 @@
+    while (true)
+    {
+      switch (*_reading_head)
+      {
+        case tokens::SQUOTE:
+          if (in_escape) break;
+          // If not in an escape sequence, terminate scanning for the name
+        case tokens::NULL_TERMINATOR:
+          goto end_name_scan;
+        case tokens::ESCAPE:
+          in_escape = !in_escape;
+      }
+
+      _reading_head++;
+    }
+    end_name_scan:
+
+    const char* token_end = _reading_head;
+
+    // Consume \'
+    read_character<tokens::SQUOTE>();
+
+    return std::string(_token_start, token_end - _token_start);
+  }
+
+  inline tensor_data read_tensor_data()
+  {
+    NO_ERROR_OR_RETURN_DEFAULT(tensor_data(bytes(), bytes()))
+
+    // Consume \'
+    read_character<tokens::SQUOTE>();
+
+    // TODO: Support reading type information for the tensor (and later map/sequence)
+    // See value_t definition in tensor_notation.h
+    // <TYPE_INFO> := '<' VALUE_TYPE '>'
+    // <VALUE_TYPE> := "float"
+    // read_value_type();
+    // read_character<';'>();
+    
+    // Read base_64(int_64t[]) until ';'
+    bytes dimensions_base64 = read_base64();
+
+    // Consume ';' - also another bit of validation that dimensions were parsed 
+    // meaningfully
+    read_character<tokens::SEMICOLON>();
+
+    // Read base_64(float[]) until '\''
+    bytes values_base64 = read_base64();
+
+    // Consume \'
+    read_character<tokens::SQUOTE>();
+
+    // Should we do the base64 parse here?
+    return std::make_pair(dimensions_base64, values_base64);
+  }
+
+  inline void read_tensor()
+  {
+    NO_ERROR_OR_RETURN();
+
+    std::string name = read_tensor_name();
+
+    skip_whitespace();
+
+    read_character<tokens::COLON>();
+
+    skip_whitespace();
+
+    tensor_data value = read_tensor_data();
+
+    _parse_context.push_input(name, value);
+  }
+
+  inline void read_tensor_list()
+  {
+    NO_ERROR_OR_RETURN();
+
+    skip_whitespace();
+
+    read_character<tokens::OPEN_CBRACKET>();
+    
+    skip_whitespace();
+
+    while (*_reading_head != tokens::CLOSE_CBRACKET 
+        && *_reading_head != tokens::NULL_TERMINATOR)
+    {
+      NO_ERROR_OR_RETURN();
+
+      read_tensor();
+      skip_whitespace();
+      if (*_reading_head == tokens::COMMA)
+      {
+        read_character<tokens::COMMA>();
+        skip_whitespace();
+      }
+    }
+
+    read_character<tokens::CLOSE_CBRACKET>();
+  }
+
+private:
+  const char* const _parse_buffer;
+  
+  const char* _token_start;
+  const char* _reading_head;
+
+  OnnxRtInputContext& _parse_context;
+
+  bool _has_error;
+  std::string _error_message;
+};
+
+TensorParser::TensorParser(const char* tensor_notation, OnnxRtInputContext& input_context) 
+: _parse_buffer{tensor_notation}, 
+  _reading_head{tensor_notation}, 
+  _token_start{tensor_notation}, 
+  _parse_context{input_context},
+  _has_error{false},
+  _error_message{""}
+{
+  if (_parse_buffer && *_parse_buffer != tokens::NULL_TERMINATOR)
+  {
+    read_tensor_list();
```

Comment:
Why start off the parser in the constructor?

Reply:
Because the semantics were "parsing can happen once" per Parser instance. This indicates a flawed abstraction, in particular that we pass the OnnxRtInputContext into the parser (to parallel what happens with VW), but it seems like a better design would be to have ORIC to drive the parser inside of its constructor, or a factory method, or similar.

For now pulled this out into `cached_parse()`. I will try to rework this a bit when I give ExampleBuilder another go.